numpy/core/src/umath/ufunc_object.c

```diff
@@ -2825,16 +2859,15 @@ reducelike_promote_and_resolve(PyUFuncObject *ufunc,
         goto fail;
     }
     /* TODO: This really should _not_ be unsafe casting (same above)! */
-    if (validate_casting(ufuncimpl,
-            ufunc, ops, out_descrs, NPY_UNSAFE_CASTING) < 0) {
+    if (validate_casting(ufuncimpl, ufunc, ops, out_descrs, casting) < 0) {
         goto fail;
     }
 
     return ufuncimpl;
 
   fail:
     for (int i = 0; i < 3; ++i) {
-        Py_DECREF(out_descrs[i]);
+        Py_CLEAR(out_descrs[i]);
```

Comment:
Is this because the input may be NULL?

Reply:
They can't be NULL here, but I want to make sure they are NULL'ed for the caller to not clean up a second time.  `Py_CLEAR` is just one way to do it (I think you are right and it also does a NULL check).